docs/adr/0008-use-tracing.md

```diff
@@ -0,0 +1,165 @@
+# Replace logging and error reporting infrastructure with tracing.
+
+* Status: proposed
+* Deciders:
+  * AppServices team: ?
+  * Firefox Android: ?
+  * Firefox iOS: ?
+  * Firefox Desktop: ?
+
+* Date: Mar 06, 2024
+* Feedback deadline: Mar 20, 2024
+
+## Context and Problem Statement
+
+Currently all Rust code in [app-services](https://github.com/mozilla/application-services/)
+uses the `log` crate for debugging and diagnostics.
+Log output is often captured during tests where it's just printed to stdout/stderr if the test
+fails. Further, both our mobile apps and Firefox Desktop have techniques for having this log
+output be sent across the FFI so it can be captured in whatever "native" logging system is
+in place (eg, logcat on Android, custom file logging on iOS, Log.sys.mjs/Console.sys.mjs logging on Desktop)
```

Comment:
I don't think this is true - getting logs from tests happens before it is exposed to consumers. Getting logs from tests when making changes happens before the code is even checked in and exposed to consumers.

To put it another way: none of our consumers currently capture trace logs without making source changes and making a custom build - but trace logging is very valuable regardless.

Reply:
(oops - I meant to say that I don't object to swapping the sentences, but was just noting that IMO it's not entirely accurate that the main point is for consumers, but that the *initial* point is tests and consumers is subsequent)

Reply:
I wonder if it's also worth capturing that we regularly ask users to send us logs for troubleshooting, and that informs how we approach logging.

I get the impression that trace-level logging for other Gecko crates is intended more for developers more than users, whereas we specifically want users to be able to easily turn it on, send us the log files, and turn it back off.

Maybe something like:

> Log output is valuable for working on the code, and for helping users troubleshoot. When working on the code, log output is often captured during tests, where it's just printed to stdout/stderr if the test fails. For helping users troubleshoot, log output is sent across the FFI so it can be captured in whatever "native" logging system is in place. Users can then export and attach the logs to a bug.

Reply:
That's a good point! I wonder if we could expand that further to reflect that unlike most Rust code in our apps, app-services is kinda "higher level", akin to Firefox JS code, where logging to the console in the expectation a motivated user can find the log is relatively normal.